Cosmos integration

wa-aal opened the pull request:
## Context (issues, jira)

First Draft PR with structured file and functions

## Description / Usage

<!-- please share a sample of code that uses your new code (if features were added) -->
<!-- please document quickly what would the "user land" need to do to use the feature -->

## Expectations

- [ ] **Test coverage: The changes of this PR are covered by test.** Unit test were added with mocks when depending on a backend/device.
- [ ] **No impact: The changes of this PR have ZERO impact on the userland.** Meaning, we can use these changes without modifying LLD/LLM at all. It will be a "noop" and the maintainers will be able to bump it without changing anything.

<!--
If one of these can't be checked, please document it carefully (on the reason you can't check it).
NB: We want to avoid as much as possible such breaking changes to make a bump very easy.
-->

alexalouit committed:
Update env.ts

new endpoint

alexalouit committed:
Create Cosmos.ts

first cosmos API version

alexalouit committed:
Create js.ts

First bridge version

alexalouit committed:
Update package.json

Add cosmjs package

alexalouit committed:
Update Cosmos.ts

separate getTransactions and getTransaction

alexalouit committed:
Update package.json

Add cosmjs package

alexalouit committed:
Update Cosmos.ts

Introduce mixed solution

alexalouit committed:
Update package.json

Add cosmjs package

alexalouit committed:
Update Cosmos.ts

Handle tendermint-rpc low level query
Fix getTransaction
Fix getTransactions
Fix getAllBalances

alexalouit committed:
Update Cosmos.ts

use common tmClient
always return transaction date

alexalouit committed:
Update js.ts

fix txToOps

alexalouit committed:
Update Cosmos.ts

fix transactions

alexalouit committed:
Update js.ts

fix txToOps
fix getAccountShape

alexalouit committed:
Update package.json

Add cosmjs package

alexalouit committed:
Update Cosmos.ts

prepare fee transaction
prepare transactions pagination

alexalouit committed:
Update Cosmos.ts

fix typo

alexalouit committed:
Update Cosmos.ts

add getFees method
remove useless method
fix some things

alexalouit committed:
Update js.ts

fix txToOps
fix amount
fix fee
remove useless method

alexalouit committed:
Update Cosmos.ts

fix default balance

WA-SerapK committed:
Pagination support

alexalouit committed:
Update Cosmos.ts

fix transactions order

alexalouit committed:
Update Cosmos.ts

fix typo

alexalouit committed:
Update js.ts

fix typo

alexalouit committed:
Update js.ts

remove useless import

alexalouit committed:
Update Cosmos.ts

fix transactions iteration

alexalouit committed:
Update Cosmos.ts

Fix typo
Support outgoing transactions
Sort transactions

alexalouit committed:
Update js.ts

Unified transaction to operation logic
Handle delegate and reward transactions

alexalouit committed:
Update js.ts

More unified transaction to operation logic

alexalouit committed:
Delegations support

Fetch delegations

alexalouit committed:
Update Cosmos.ts

fix transaction fee amount

alexalouit committed:
Update js.ts

fix transaction balance

alexalouit committed:
Update js.ts

bugfix

alexalouit committed:
Update Cosmos.ts

bugfix

alexalouit committed:
Update Cosmos.ts

fix getAllBalances return value

alexalouit committed:
Update js.ts

support withdraw address

alexalouit committed:
Update js.ts

Fix delegatedBalance, pendingRewardsBalance and unbondingBalance

alexalouit committed:
Update js.ts

fix balance

alexalouit committed:
Update Cosmos.ts

fix status validators

alexalouit committed:
Update Cosmos.ts

fix DecodedTxRaw and decodeTxRaw import

alexalouit committed:
Update js.ts

fix balance
fix typo

alexalouit committed:
Update js.ts

fix import

alexalouit committed:
Update js.ts

remove useless import

alexalouit committed:
Update js.ts

postSync code reduction

alexalouit committed:
Update js.ts

updateTransaction code reduction

alexalouit committed:
Update js.ts

prepareTransaction code reduction

alexalouit committed:
Update js.ts

remove useless code

alexalouit committed:
Update js.ts

estimateMaxSpendable code reduction

alexalouit committed:
Update Cosmos.ts

remove useless import

alexalouit committed:
Update js.ts

more accurate operation senders/recipients (handle duplicates)

alexalouit committed:
Update js.ts

fix typo

alexalouit committed:
new getAccount method

More unified getAccount method

alexalouit committed:
Update Cosmos.ts

remove useless import

alexalouit committed:
Update js.ts

remove useless import

alexalouit committed:
Update js.ts

first attempt to sign with hardware

alexalouit committed:
Update package.json

add cosmjs package

alexalouit committed:
Update Cosmos.ts

remove useless dependencies

alexalouit committed:
Update js.ts

remove useless dependencies

alexalouit committed:
Update Cosmos.ts

fix getChainId

alexalouit committed:
Update Cosmos.ts

remove useless sign api method

alexalouit committed:
Update Cosmos.ts

remove useless signAndBroadcast api method

alexalouit committed:
Update Cosmos.ts

add getSequence api method

alexalouit committed:
Update Cosmos.ts

fix brodcast

alexalouit committed:
Update js.ts

fix sign and broadcast

WA-SerapK committed:
Update Cosmos.ts

Add isValidRecipient method

alexalouit committed:
Update Cosmos.ts

fix typo

alexalouit committed:
getAccount improvement

more unified getAccount information
signOperation now support chainId, sequence, and accountNumber

alexalouit committed:
Update js.ts

Add isValidRecipent process

alexalouit committed:
Update Cosmos.ts

fix getAccount return value (when wallet is empty)

alexalouit committed:
Update js.ts

fix currency unit

alexalouit committed:
improvements

reduce dependencies
use direct API call

alexalouit committed:
Update js.ts

prepare to broadcast transaction

alexalouit committed:
Update package.json

update cosmjs version

alexalouit committed:
Update package.json

update package

alexalouit committed:
fix api 404 error

handle trailing slash

alexalouit committed:
fix broadcast

alexalouit committed:
Update js.ts

adjust query order

haammar-ledger commented on src/api/Cosmos.ts:
```diff
@@ -0,0 +1,376 @@
+import { patchOperationWithHash } from "../operation";
+import { calculateFee, GasPrice } from "@cosmjs/stargate";
+
+let tmClient;
```
Since we're only using http (no websocket), I think you can instantiate the client only once, for example:
```
let tmClient;
const getTendermint34Client = async () => {
  if (!tmClient) {
    tmClient = await Tendermint34Client.connect(defaultRpcEndpoint);
  }
  return tmClient;
};
```

haammar-ledger replied:
Also well done removing all other clients, if they were not needed.
To be checked if the Tendermint one can be removed too maybe?

wa-aal replied:
fix in 869fbfb4

haammar-ledger reviewed (commented):
I reviewed the API so far, will review the rest now.

Some general comments:
- Putting the network API in `src/api` is deprecated, please move `Cosmos.ts` to a `src/families/cosmos/api` folder.
- Please split the `js.ts` file following the convention described in the developer portal: https://developers.ledger.com/docs/coin/live-common/#split-your-code it will help readability and maintenance.

haammar-ledger reviewed (commented):
Sync looks much better, but there may still be issues, you can check with an address containing a lot of transactions, e.g:
`ledger-live sync -c cosmos --id cosmos14lultfckehtszvzw4ehu0apvsr77afvyhgqhwh`
vs
[https://www.mintscan.io/cosmos/account/cosmos14lultfckehtszvzw4ehu0apvsr77afvyhgqhwh](https://www.mintscan.io/cosmos/account/cosmos14lultfckehtszvzw4ehu0apvsr77afvyhgqhwh)

alexalouit committed:
Update Cosmos.ts

add common promise

alexalouit committed:
Update js.ts

use built-in operation id encoder

wa-aal commented on src/api/Cosmos.ts:
```diff
@@ -0,0 +1,323 @@
+    }
+
+    // sort transactions by date
+    data.sort((a, b) => {
+      return b.date.getTime() - a.date.getTime();
+    });
```
So left unsorted in the live-common is not a big deal?
They are retrieved in a disparate way because requests for sent and received transaction are different.

wa-aal replied:
fix in 1233e6b

wa-aal commented on src/api/Cosmos.ts:
```diff
@@ -0,0 +1,323 @@
+  const blockHeight = await getHeight();
+  const txs = await getTransactions(address);
+  const delegations = await getDelegators(address);
+  const withdrawAddress = await getWithdrawAddress(address);
```
fix in 72ac7f6c

wa-aal commented on src/families/cosmos/bridge/js.ts:
```diff
@@ -0,0 +1,697 @@
+      return op.senders.indexOf(element) === index;
+    });
+
+    op.id = `${id}-${hash}-${op.type}`;
```
fix in 48effb3f

wa-aal commented on src/api/Cosmos.ts:
```diff
@@ -0,0 +1,323 @@
+    });
+
+    for (const tx of data) {
+      txs[tx.hash] = tx;
```
This is a way to deal with duplicates (possible collisions on pagination process). 
Maybe it is already handled somewhere else?

wa-aal replied:
fix in 1233e6b1

haammar-ledger commented on src/api/Cosmos.ts:
```diff
@@ -0,0 +1,385 @@
+
+  const { data } = await network({
+    method: "POST",
+    url: `${defaultEndpoint}/txs`,
```
As mentioned by e-mail, please try with the correct endpoint for new transactions format: `/cosmos/tx/v1beta1/txs`

wa-aal replied:
fix in 22a8559c

alexalouit committed:
Update js.ts

useless transaction filtering

haammar-ledger commented on src/families/cosmos/bridge/js.ts:
```diff
@@ -0,0 +1,738 @@
+        const { accountNumber, sequence } = await getAccount(freshAddress);
+        const chainId = await getChainId();
+
+        const unsigned = {
+          chain_id: chainId,
+          account_number: accountNumber,
+          sequence: sequence,
+          fee: fee,
+          msgs: msgs,
+          memo: transaction.memo || "",
+        };
```
You can extract the build of the unsigned payload in a dedicated `buildTransaction` function, that may be also used to compute fees.

wa-aal replied:
fix 94d5919

haammar-ledger reviewed (commented):
Second pass, on the bridges.
General comments from testing:

- Signing : sometimes OK but I have a glitch in some case:
```
$ ledger-live send -c cosmos -i 1 --amount 0.001 --recipient cosmos1g84934jpu3v5de5yqukkkhxmcvsw3u2ajxvpdl
→ FROM Cosmos 2: 0.01005 ATOM (6ops) (cosmos108uy5q9jt59gwugq5yrdhkzcd9jryslmpcstk5 on 44'/118'/1'/0/0) #1 js:2:cosmos:cosmos108uy5q9jt59gwugq5yrdhkzcd9jryslmpcstk5: (! sum of ops 0.01 ATOM) 0.01005 ATOM spendable. 
✔️ transaction 
SEND  0.001 ATOM
TO cosmos1g84934jpu3v5de5yqukkkhxmcvsw3u2ajxvpdl

with fees=0.002
STATUS 
  amount: 0.001 ATOM
  estimated fees: 0.002 ATOM
  total spent: 0.003 ATOM
Address cosmos108uy5q9jt59gwugq5yrdhkzcd9jryslmpcstk5 not found in wallet
```

- Broadcast : beyond using the right URL, it seems the payload doesn't look like what is expected.
  - doc: [https://docs.cosmos.network/master/run-node/txs.html#broadcasting-a-transaction-4](https://docs.cosmos.network/master/run-node/txs.html#broadcasting-a-transaction-4)
  - currently:
`http: post https://cosmoshub4.coin.ledger.com/cosmos/tx/v1beta1/txs {"data":{"tx":{"chain_id":"cosmoshub-4","msg":[{"type":"cosmos-sdk/MsgSend","value":{"from_address":"cosmos1g84934jpu3v5de5yqukkkhxmcvsw3u2ajxvpdl","to_address":"cosmos108uy5q9jt59gwugq5yrdhkzcd9jryslmpcstk5","amount":[{"denom":"atom","amount":"0.001"}]}}],"fee":{"amount":[{"amount":"0.002","denom":"atom"}],"gas":"null"},"signatures":[{"pub_key":{"type":"tendermint/PubKeySecp256k1","value":"A4hFmyZTUZlIsSSS8aC0ZHIBEMFHqBVdI9Qjpcw8Idia"},"signature":"HUegjrcAILqWeo66eWqWR+DIzyaOVbZDkj6c2eu76gUHbFPu7LO0Kt4LqCAfbBXflXC8ujy0VchTUb5PcYkDPA==","sequence":"70","account_number":"31094"}],"memo":""},"mode":"commit"}}`

alexalouit committed:
Update Cosmos.ts

fix legacy endpoint

alexalouit committed:
restruct files

split files as "js-" format
move cosmos api file location from legacy directory to new

alexalouit committed:
mergeOps support

let mergeOps handle sorting and dedup

alexalouit committed:
Update js-signOperation.ts

use built-in operation id encoder

alexalouit committed:
Update js-signOperation.ts

move from cosmjs amino signer to cosmos app

alexalouit committed:
Update Cosmos.ts

remove useless returned data
set return type as strict

alexalouit committed:
Update Cosmos.ts

move to new broadcast api endpoint

alexalouit committed:
tendermint endpoint

temporary use another tendermint endpoint

alexalouit committed:
Update Cosmos.ts

remove useless import

alexalouit committed:
Update Cosmos.ts

remove useless endpoint

alexalouit committed:
Update js-signOperation.ts

fix typo

alexalouit committed:
Update js-signOperation.ts

prepare delegate payload

alexalouit committed:
Update js-signOperation.ts

prepare undelegate payload

alexalouit committed:
prepareTransaction

add prepareTransaction process
preparation of fees calculation

alexalouit committed:
Update package.json

add cosmjs package

alexalouit committed:
Update Cosmos.ts

fix broadcast

alexalouit committed:
Update js-signOperation.ts

fix sign transaction

alexalouit committed:
instruct input data to provider

alexalouit committed:
Update Cosmos.ts

add simulation api endpoint

alexalouit committed:
Update js-signOperation.ts

remove useless dependencies

alexalouit committed:
Update js-prepareTransaction.ts

preprare payload

alexalouit committed:
Update js.ts

fix typo

alexalouit committed:
Create js-buildTransaction.ts

preprare payload

alexalouit committed:
Update js-signOperation.ts

fix signature
fix derivation
use payload

wa-aal commented on src/families/cosmos/bridge/js.ts:
```diff
@@ -0,0 +1,738 @@
+  });
+};
+
+const prepareTransaction = async (a, t) => t;
```
fix in 94d5919

alexalouit committed:
Update js-updateTransaction.ts

more complex updateTransaction

wa-aal commented on src/families/cosmos/bridge/js.ts:
```diff
@@ -0,0 +1,738 @@
+  validators: [],
+});
+
+const updateTransaction = (t, patch) => ({ ...t, ...patch });
```
fix in abd7953f

alexalouit committed:
getTransactions

use native network call

alexalouit committed:
Update Cosmos.ts

remove useless import

wa-aal commented on src/api/Cosmos.ts:
```diff
@@ -0,0 +1,323 @@
+    const txsInNbPage = Math.ceil((txsIn.totalCount - txsInDone) / perPage);
+
+    // paginate over pages
+    for (let i = 2; i <= txsInNbPage; i++) {
```
fix in 869fbfb4

wa-aal commented on src/env.ts:
```diff
@@ -55,6 +55,11 @@ const envDefinitions = {
     desc: "Node API to use for cosmos (COSMOS_NODE or STARGATE_NODE are known)",
   },
+  API_COSMOS_RPC_URL: {
+    def: "https://rpc.cosmos.network/",
```
fix in c886569a

alexalouit committed:
Update js-signOperation.ts

fix note

alexalouit committed:
Update js-signOperation.ts

fix typo

alexalouit committed:
Update js-signOperation.ts

fix signature

alexalouit committed:
Update js-signOperation.ts

remove useless dependency

alexalouit committed:
Update js-signOperation.ts

fix typo

alexalouit committed:
Update js-signOperation.ts

fix amount value

alexalouit committed:
Update js-prepareTransaction.ts

add default memo

alexalouit committed:
Update js-signOperation.ts

fix balance

alexalouit committed:
Update js-signOperation.ts

fix typo

alexalouit committed:
use currency code

alexalouit committed:
Update js-buildTransaction.ts

use libcore default gas

alexalouit committed:
Update js-signOperation.ts

remove useless

alexalouit committed:
Update js-estimateMaxSpendable.ts

fix estimateMaxSpendable

alexalouit committed:
Update Cosmos.ts

simulate must be silent

alexalouit committed:
Update js-prepareTransaction.ts

fix cosmos new estimation method

alexalouit committed:
Update js-prepareTransaction.ts

fix return value when simulate fail

wa-aal commented on src/families/cosmos/bridge/js.ts:
```diff
@@ -0,0 +1,738 @@
+  return Promise.resolve(
+    BigNumber.max(0, mainAccount.balance.minus(estimatedFees))
+  );
+};
```
fix in 0422023b and c478e167

wa-aal commented on src/api/Cosmos.ts:
```diff
@@ -0,0 +1,323 @@
+    api = await StargateClient.connect(defaultRpcEndpoint);
+    data = await api.getAccount(address);
+    // eslint-disable-next-line no-empty
+  } catch (e) {}
```
fix, now using network call

gre commented on package.json:
```diff
@@ -42,6 +42,11 @@
+    "@cosmjs/ledger-amino": "^0.26.5",
+    "@cosmjs/proto-signing": "^0.26.5",
+    "@cosmjs/stargate": "^0.26.5",
+    "@cosmjs/tendermint-rpc": "^0.26.5",
```
please commit the yarn.lock file as well

alexalouit committed:
Update Cosmos.ts

return merged responses instead of array

alexalouit committed:
Update js-synchronisation.ts

more readable operating logic

alexalouit committed:
fix payload

alexalouit committed:
Update js-synchronisation.ts

estimatedFees can't be calculate without account

alexalouit committed:
packages

remove useless package
fix yarn lockfile

Commit:
Merge branch 'master' into dev

alexalouit committed:
simulate: more strict input type

alexalouit committed:
Update js-synchronisation.ts

handle cosmospub address

alexalouit committed:
add redelegations support

alexalouit committed:
add unbondings support

alexalouit committed:
Merge branch 'dev' of github.com:wa-aal/ledger-live-common into dev

alexalouit committed:
ci adjustments

add cosmos and js to dataset test
define cosmos as experimental integration
remove useless test scripts

alexalouit committed:
disable test in ci

alexalouit committed:
Update js-prepareTransaction.ts

use existing var

alexalouit committed:
Update js-getTransactionStatus.ts

bugfix

haammar-ledger commented on src/families/cosmos/js-prepareTransaction.ts:
```diff
@@ -0,0 +1,78 @@
+
+  transaction.gas = new BigNumber(
+    simulation?.gas_info?.gas_used || 0
+  ).multipliedBy(gasPrice);
+
+  transaction.fees = gasPrice
+    .multipliedBy(transaction.gas)
```
It seems this multiplies gasPrice twice? I mean it does :
`transaction.fees = gasPrice.multipliedBy(gas_used.multipliedBy(gasPrice));`
?

wa-aal replied:
fix in 9a853181

alexalouit committed:
Update Cosmos.ts

fix pendingRewards

alexalouit committed:
Update env.ts

reduce gas amplifier

alexalouit committed:
Update js-prepareTransaction.ts

fix fees calculation

haammar-ledger commented on src/families/cosmos/bridge/js.ts:
```diff
@@ -0,0 +1,72 @@
+  sync,
+  receive,
+  signOperation,
+  broadcast: async ({ signedOperation }) => {
+    return broadcast({
+      signedOperation,
+    });
+  },
```
Is this syntax necessary, can't just export `broadcast`?

wa-aal replied:
fix in 3bd2b9f0

haammar-ledger reviewed (changes requested):
- Now that cosmos is set as an experimental JS impl, you can push the generated file `src/generated/bridge/js.ts`.

- In the `cli/tests` part, was it not possible to remove the `cli/scripts/tests.sh`?
Also I don't get why the `apdu.snapshot.log` and `mandatory` files appear in this PR?

alexalouit committed:
Update test-dataset.ts

remove useless issue tag

alexalouit committed:
Update libcore-buildTransaction.ts

don't break current libcore implementation

alexalouit committed:
remove useless complex code

alexalouit committed:
restore apdu.snapshot.log and mandatory

wa-aal commented on src/env.ts:
```diff
@@ -141,7 +141,7 @@ const envDefinitions = {
   },
   COSMOS_GAS_AMPLIFIER: {
-    def: 4,
+    def: 1.4,
```
fix in 4a52b5a5

wa-aal commented on src/families/cosmos/test-dataset.ts:
```diff
@@ -455,9 +455,9 @@ const cosmos: CurrenciesData<Transaction> = {
+  implementations: ["js"],
   currencies: {
-    // cosmos, // LL-7872
+    cosmos, // LL-7872
```
fix in 730925f6

alexalouit committed:
Update js.ts

push the generated file experimental JS implementation

alexalouit committed:
Merge remote-tracking branch 'upstream/master' into dev

alexalouit committed:
fix yarn.lock
